src/wp-includes/class-wp-modules.php

```diff
@@ -0,0 +1,246 @@
+		}
+		$key = array_search( $module_identifier, $this->enqueued_before_registered, true );
+		if ( false !== $key ) {
+			array_splice( $this->enqueued_before_registered, $key, 1 );
+		}
+	}
+
+	/**
+	 * Returns the import map array.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @return array Array with an `imports` key mapping to an array of module identifiers and their respective URLs, including the version query.
+	 */
+	public function get_import_map() {
+		$imports = array();
+		foreach ( $this->get_dependencies( array_keys( $this->get_enqueued() ) ) as $module_identifier => $module ) {
+			$imports[ $module_identifier ] = $module['src'] . $this->get_version_query_string( $module['version'] );
+		}
+		return array( 'imports' => $imports );
+	}
+
+	/**
+	 * Prints the import map using a script tag with a type="importmap" attribute.
+	 *
+	 * @since 6.5.0
+	 */
+	public function print_import_map() {
+		$import_map = $this->get_import_map();
+		if ( ! empty( $import_map['imports'] ) ) {
+			wp_print_inline_script_tag(
+				wp_json_encode( $import_map, JSON_HEX_TAG | JSON_HEX_AMP ),
+				array(
+					'type' => 'importmap',
+				)
+			);
+		}
+	}
+
+	/**
+	 * Prints all the enqueued modules using script tags with type="module"
+	 * attributes.
+	 *
+	 * @since 6.5.0
+	 */
+	public function print_enqueued_modules() {
+		foreach ( $this->get_enqueued() as $module_identifier => $module ) {
+			wp_print_script_tag(
+				array(
+					'type' => 'module',
+					'src'  => $module['src'] . $this->get_version_query_string( $module['version'] ),
+					'id'   => $module_identifier,
+				)
+			);
+		}
+	}
+
+	/**
+	 * Prints the the static dependencies of the enqueued modules using link tags
+	 * with rel="modulepreload" attributes.
+	 *
+	 * If a module has already been enqueued, it will not be preloaded.
+	 *
+	 * @since 6.5.0
+	 */
+	public function print_module_preloads() {
+		foreach ( $this->get_dependencies( array_keys( $this->get_enqueued() ), array( 'static' ) ) as $module_identifier => $module ) {
+			if ( true !== $module['enqueued'] ) {
+				echo sprintf(
+					'<link rel="modulepreload" href="%s" id="%s">',
+					esc_attr( $module['src'] . $this->get_version_query_string( $module['version'] ) ),
+					esc_attr( $module_identifier )
+				);
+			}
+		}
+	}
+
+	/**
+	 * Gets the version of a module.
+	 *
+	 * If SCRIPT_DEBUG is true, the version is the current timestamp. If $version
+	 * is set to false, the version number is the currently installed WordPress
+	 * version. If $version is set to null, no version is added. Otherwise, the
+	 * string passed in $version is used.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string|false|null $version The version of the module.
+	 * @return string A string with the version, prepended by `?ver=`, or an empty string if there is no version.
+	 */
+	private function get_version_query_string( $version ) {
+		if ( defined( 'SCRIPT_DEBUG ' ) && SCRIPT_DEBUG ) {
+			return '?ver=' . time();
```

Comment:
Hmm, this isn't a thing in the regular script API. While I'm not necessarily opposed to it, I'm a bit wary of introducing it without a bit more discussion. For example:
* If this is a good idea, then it should also be done in the regular `WP_Scripts` class.
* Maybe there are other concerns though, e.g. related to sites that use `SCRIPT_DEBUG` in production. That's certainly not a good idea, but for sites that do, this behavior would make their performance a lot worse by avoiding caching.

Again, I think it may be a good idea, but I think we should validate that first with other stakeholders. Maybe leave it out for the initial iteration?

Reply:
I agree, let's leave it out for the initial version. We can add it later or keep it on higher-level abstractions, [like what Gutenberg does](https://github.com/WordPress/gutenberg/blob/e8b0a4c459dd2dd8f6775f312128cc6e7106a6bc/lib/blocks.php#L250).

Removed in https://github.com/WordPress/wordpress-develop/pull/5818/commits/fda294d858731120c1651dbec70aa3468030f726.